va/va.go

```diff
@@ -678,6 +690,63 @@ func (va *ValidationAuthorityImpl) validateTLSSNI01(ctx context.Context, identif
 	return va.validateTLSSNI01WithZName(ctx, identifier, challenge, ZName)
 }
 
+func (va *ValidationAuthorityImpl) validateTLSALPN01(ctx context.Context, identifier core.AcmeIdentifier, challenge core.Challenge) ([]core.ValidationRecord, *probs.ProblemDetails) {
+	if identifier.Type != "dns" {
+		va.log.Info(fmt.Sprintf("Identifier type for TLS-ALPN-01 was not DNS: %s", identifier))
+		return nil, probs.Malformed("Identifier type for TLS-ALPN-01 was not DNS")
+	}
+
+	certs, cs, validationRecords, problem := va.tryGetTLSCerts(ctx, identifier, challenge, &tls.Config{
+		NextProtos: []string{ACMETLS1Protocol},
+		ServerName: identifier.Value,
+	})
+	if problem != nil {
+		return validationRecords, problem
+	}
+
+	if !cs.NegotiatedProtocolIsMutual || cs.NegotiatedProtocol != ACMETLS1Protocol {
+		errText := fmt.Sprintf(
+			"Cannot negotiate ALPN protocol %q for %s challenge",
+			ACMETLS1Protocol,
+			core.ChallengeTypeTLSALPN01,
+		)
+		return validationRecords, probs.Unauthorized(errText)
+	}
+
+	leafCert := certs[0]
+
+	// Verify SNI - certificate returned must be issued only for the domain we are verifying.
+	if len(leafCert.DNSNames) != 1 || !strings.EqualFold(leafCert.DNSNames[0], identifier.Value) {
+		hostPort := net.JoinHostPort(validationRecords[0].AddressUsed.String(), validationRecords[0].Port)
+		names := certNames(leafCert)
+		errText := fmt.Sprintf(
+			"Incorrect validation certificate for %s challenge. "+
+				"Requested %s from %s. Received %d certificate(s), "+
+				"first certificate had names %q",
+			challenge.Type, identifier.Value, hostPort, len(certs), strings.Join(names, ", "))
+		return validationRecords, probs.Unauthorized(errText)
+	}
+
+	// Verify key authorization in acmeValidation extension
+	h := sha256.Sum256([]byte(challenge.ProvidedKeyAuthorization))
+	for _, ext := range leafCert.Extensions {
+		if IdPeAcmeIdentifierV1.Equal(ext.Id) && ext.Critical {
```

Comment:
Might want to add a specific error if the extension is present but non-critical, otherwise we fall out and return a error that indicates there was no extension at all.